sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -80,28 +105,57 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
    * transformExpressionsDown or transformExpressionsUp should be used.
    *
    * @param rule the rule to be applied to every expression in this operator.
+   * @param cond  a Lambda expression to prune tree traversals. If `cond.apply` returns false
+   *              on an expression T, skips processing T and its subtree; otherwise, processes
+   *              T and its subtree recursively.
+   * @param ruleId is a unique Id for `rule` to prune unnecessary tree traversals. When it is
+   *        RuleId.UnknownId, no pruning happens. Otherwise, if `rule`(with id `ruleId`) has been
+   *        marked as in effective on an expression T, skips processing T and its subtree. Do not
+   *        pass it if the rule is not purely functional and reads a varying initial state for
+   *        different invocations.
    */
-  def transformExpressions(rule: PartialFunction[Expression, Expression]): this.type = {
-    transformExpressionsDown(rule)
+  def transformExpressions(rule: PartialFunction[Expression, Expression],
+    cond: TreePatternBits => Boolean = AlwaysProcess.fn,
```

Comment:
Can we have helpers for this? It is a bit annoying to write place holder syntax all the time. I think the most common one should be contains(..) right?

Reply:
>> one should be contains(..)

I think there're three variations:
- containsPattern(P)
- containsAllPatterns(P1, P2, P3)
- containsAnyPatterns(P1, P2, P3)

If we have a helper, we still need to pass the helper with (1) pattern enums and (2) an "all" v.s. "any" enum. 
We can discuss offline later.

Reply:
I checked the prototype implementation and it turns out that there're multiple variations, e.g.,

- containsPattern(P)
- containsAllPatterns(P1, P2, P3)
- containsAnyPattern(P1, P2, P3)
- containsAllPatterns(P1, P2) && containsAnyPattern(P3, P4)
- containsAllPatterns(P1, P2) && !containsAnyPattern(P3, P4)
- containsAllPatterns(P1, P2) || containsAllPatterns(P3, P4)

Thus, I kept it as it is. Let me know if you have any further thought.